grant-project/grant-monorepo
# Cancel / Refund proposal

### What This Does

This gives trustees the ability to cancel ongoing proposals, or refund successfully funded proposals. It also adds a bit of UI where an edit button will live when we implement that.

This also adds handling for a lot more refund states. Contributors can now refund if a project fails to reach its funding target before the deadline, or if the project was canceled by a trustee.

### Steps to Test

* Create a proposal with account 1, and make that account a trustee
* Partially fund the proposal with account 2
* Back to account 1, cancel the proposal from the actions dropdown
* As account 2, refund the proposal. Confirm it all worked and you got your money back, minus gas fees.
* Do the same, but fully funding the proposal as account 2.
* Do the same, but allowing the proposal to hit its deadline with partial funding.
* Make sure that refund voting still works correctly.

## Screenshots

### New actions dropdown visible only by trustees

<img width="300" alt="screen shot 2018-09-24 at 3 45 38 pm" src="https://user-images.githubusercontent.com/649992/46033182-8cec7180-c0cb-11e8-9a2b-a1a82e9f35b8.png">


### Refund modal (copy is dynamic based on proposal state)

<img width="538" alt="screen shot 2018-09-24 at 3 48 25 pm" src="https://user-images.githubusercontent.com/649992/46033125-61698700-c0cb-11e8-8c8a-4466f94362f4.png">

### New placeholder states for governance

<img width="1318" alt="screen shot 2018-09-24 at 4 14 59 pm" src="https://user-images.githubusercontent.com/649992/46033132-69292b80-c0cb-11e8-82f1-175cc95b33fe.png">

### Governance state if proposal is canceled before funding

<img width="1306" alt="screen shot 2018-09-24 at 4 32 17 pm" src="https://user-images.githubusercontent.com/649992/46033147-747c5700-c0cb-11e8-9ac9-9c3dbdf55e26.png">
